framework/include/cppmicroservices/BundleContext.h

```diff
@@ -607,15 +609,169 @@ class US_Framework_EXPORT BundleContext
     return ServiceObjects<S>(d, reference);
   }
 
-  void AddServiceListener(const ServiceListener& delegate,
-                          const std::string& filter = std::string());
-  void RemoveServiceListener(const ServiceListener& delegate);
+  /**
+   * Adds the specified <code>callable</code> with the
+   * specified <code>filter</code> to the context bundles's list of listeners.
+   * See LDAPFilter for a description of the filter syntax. Listeners
+   * are notified when a service has a lifecycle state change.
+   *
+   * <p>
+   * You must take care to remove registered listeners before the <code>callable</code>
+   * object is destroyed. However, the Micro Services framework takes care
+   * of removing all listeners registered by this context bundle's classes
+   * after the bundle is stopped.
+   *
+   * <p>
+   * The callable is called if the filter criteria is met. To filter based
+   * upon the class of the service, the filter should reference the
+   * Constants#OBJECTCLASS property. If <code>filter</code> is
+   * empty, all services are considered to match the filter.
+   *
+   * <p>
+   * When using a <code>filter</code>, it is possible that the
+   * <code>ServiceEvent</code>s for the complete lifecycle of a service
+   * will not be delivered to the callable. For example, if the
+   * <code>filter</code> only matches when the property <code>x</code> has
+   * the value <code>1</code>, the callable will not be called if the
+   * service is registered with the property <code>x</code> not set to the
+   * value <code>1</code>. Subsequently, when the service is modified
+   * setting property <code>x</code> to the value <code>1</code>, the
+   * filter will match and the callable will be called with a
+   * <code>ServiceEvent</code> of type <code>SERVICE_MODIFIED</code>. Thus, the
+   * callable will not be called with a <code>ServiceEvent</code> of type
+   * <code>SERVICE_REGISTERED</code>.
+   *
+   * @param callable Any type of callable.
+   * @param filter The filter criteria.
+   * @returns a ListenerToken object which can be used to remove the callable from the
```

Comment:
Jeff, it looks like "callable object" is the best word to describe the type of listeners which can be passed to the function. So, it makes more sense not to name the function parameter as "callable" and I'll name it something appropriate and less confusing.